---
title: COMP90014 Algorithms for Functional Genomics, Semester 2 - 2016, Assignment
  3
author: "Lachlan Dryburgh 188607"
date: "7/10/2016"
output: pdf_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Libraries used for analysis
We need to load some R libraries for this differential gene expression analysis.  The main one is the edgeR package.  If they are not already installed edgeR can be installed with bioconductor the other packages can be installed with install.packages

```{r libraries, message = FALSE}
library(edgeR)
library(locfit)
library(gridExtra)
library(grid)
```
# Task 1
## Load data and create DGE list
Load the counts data.  The data is in a tab delimited file.
The counts data can then be used to create a DGEList.  The first 15 columns in the counts are females the next 15 are Males.

```{r load}
counts = read.delim("assignment3data.tsv")
d=DGEList(counts=counts[,1:30], group=factor(c(rep("F", 15),rep("M",15))))
```

## Test the difference in counts
Add a normalisation factor to the differential expression comparisons and calculate the dispersion.  Then we can perform a test between the counts.

```{r Test}
d=calcNormFactors(d)
d=estimateDisp(d)

et=exactTest(d)
```

## Top 10 differentially expressed genes
Find the top 10 differentially expressed genes based on the Benjamini-Hochberg adjsted p-value usng the topTags function.  We can add the gene names and chromosomes back into the top tags table from the count data and then use the grid library to produce a nice table.
```{r Top 10 genes}
tt = topTags(et, n=10, adjust.method="BH", sort.by="P")
tt$table=round(tt$table,4)
tt$table["Gene"]=counts[rownames(tt),][31]
tt$table["Chr"]=counts[rownames(tt),][32]
grid.table(tt$table)
```


We can see all the genes come from the Y chromosome except the Xist gene which is expressed in females to inactvate one of the copies copies the X chromosome.


## Heatmap of top 15 genes
The top tags can be used to generate a table of the top 15 differentially expressed genes, again using the BH adjusted p-value.  We can visuallise these genes along with the expression level in each individual using a heatmap.

```{r}
de = topTags(et, n=15, adjust.method="BH", sort.by="P")

heatmap(d$counts[rownames(d) %in% rownames(de),])
```

We can see on the heat map that 14 of the top 15 have expression only in the males with the remaining gene being expressed only in the females.

## Number of differentially expressed genes
We can use the exactTest function to calculate the p-value for each gene, that is the probability of seeing that level of differential expression by chance where there was no underlying difference between the groups.  A p-value cutoff of 0.05 is used.

```{r Unadjusted p-values}
ua =decideTestsDGE(et, adjust.method = "none",p.value = 0.05)

summary(ua)
sum(ua!=0)
sum(ua!=0)/length(ua)
```

We can see the number the number of differentially expressed genes is 864 which 0.0196 of the total number of genes.  However since we are calculating the p-value independently for each of 44061 genes using the unadjusted 0.05 p-value will produce many false postive results.  We can take into acount that we are performing many test and calculate an adjusted p-value using the Benjamini-Hochberg (BH) method.

```{r Adjusted p-value}
bh = decideTestsDGE(et, adjust.method = "BH", p.value = 0.05)

summary(bh)
sum(bh!=0)
sum(bh!=0)/length(bh)
```

With the adjusted p-values we can now see we have just 61 differentially expressed genes 0.0014 of the total number of genes.

## Y Chromosome genes
We want investigate how many of the genes on the Y chromosome are differentially expressed.

```{r Y Chromosome}
sum(bh[counts[32]=='Y'])
sum(counts[32]=='Y')
```

Using the 'BH' adjusted p-values we get differential expression of 24 of 115 genes on the Y chromsome.

## XIST Gene
The XIST gene is expressed to silince one the copies of the X chromosome in females.  Looking at the expression of this gene.

```{r XIST Gene}
et$table[counts[31]=="XIST",]
```

We can see a log Fold change of -9.7608 for this gene, showing that it has much lower expression in the males than the females as expected.

# Task 2
## Generate random groups
We will be comparing our results to the case where individuals are randomlly assigned to each group rather than divided by sex.  Using the sample function with after setting a seed will allow reproducable results.

```{r Sample}
set.seed(1)
random.group = sample(c(rep("F",15),rep("M",15)), size = 30, replace = F)
random.group
```

## Differential Gene Expression Analysis
We perform the same analysis as task 1 on our randomlly assigned grouping.

```{r DGE Analysis}
random.dge=DGEList(counts=counts[,1:30], group=factor(random.group))
random.dge=calcNormFactors(random.dge)
random.dge=estimateDisp(random.dge)
r.et=exactTest(random.dge)
```

## Random Top 10 
The top 10 differentially expressed genes calculated the same way as in task 1

```{r Top 10}
r.tt = topTags(r.et, n=10, adjust.method="BH", sort.by="P")
r.tt$table=round(r.tt$table,4)
r.tt$table["Gene"]=counts[rownames(r.tt),][31]
r.tt$table["Chr"]=counts[rownames(r.tt),][32]
grid.table(r.tt$table)
```

## Random Heatmap
This is the heatmap of the random group assignment data.

```{r Random heatmap}
r.de = topTags(r.et, n=15, adjust.method="BH", sort.by="P")

heatmap(random.dge$counts[rownames(random.dge) %in% rownames(r.de),])
```

We can see here that when randomally assigned the gtop differentially expressed genes are far less cleanly seperated by the group.

## Number of differentially expressed genes
Counting the number of genes with an unadjusted p-value of less than 0.05 using the same method as task 1.

```{r random unadjusted}
r.ua =decideTestsDGE(r.et, adjust.method = "none",p.value = 0.05)
summary(r.ua)
sum(r.ua!=0)
sum(r.ua!=0)/length(r.ua)
```
The number of genes with an unadjusted p-value is 1040 which is 0.0236 of the total number of genes.

```{r random adjusted}
r.bh = decideTestsDGE(r.et, adjust.method = "BH", p.value = 0.05)
summary(r.bh)
sum(r.bh!=0)
sum(r.bh!=0)/length(r.bh)
```

However when we use the Benjamini-Hochberg method to adjust the p-value we only have 1 gene with a value under 0.05 which is less than 0.0001 of the total number of genes.

## Y Chromosome
```{r random Y}
sum(r.bh[counts[32]=='Y'])
sum(counts[32]=='Y')
```
 When randomly assigned groups are assigned there are no differentially expressed genes from the Y chromosome.
 
## XIST Gene
```{r Random XIST}
r.et$table[counts[31]=="XIST",]
```
The XIST gene has a log fold change of -0.4047 which is a small change.

#Task 3

## Differential Expression
When comparing men to women it was found that 9 of the top 10 differentially expressed genes were on the Y chromosome.  Furthermore of the genes with and adjusted p-value less tan 0.05 24 out 61 where from the Y chromosome and this was from only 115 genes in total on the Y chromosome of 44061 gene in total.  This was an expected result since females do not a Y chromosome they must have zero expression of those genes and therefore any gene on the Y chromosome with significant expression will be found to be differentially expressed.  This could be seen on the heatmap representation of the top 15 genes where 14 were from the Y chromosome and we could see an very clean separation between the groups.  

Only a single gene from the X chromosome was in the top ten differentially expressed genes this is because despite females having two copies of the chromosome compared to males one off the copies is "silenced" and does not express most of it's genes.  However the XIST gene is expressed by only a silenced X chromosome therefore it will not be expressed in males since there single copy of the X chromosome is never silenced therefore it was also found in the list of differentially expressed genes. Male and females have equal numbers of the remaining other chromosomes, there is some differential expression of some gene through signaling and epigenetics however we would expect this to be far less common than on the Y chromosome and this was reflected by the fact that only 37 of the 43946 genes not on the Y chromosome were differentially expressed.


## Randomised Control
In order to validate our analysis we repeated the analysis with individuals randomly assigned to groups.  If we found differential expression between randomly assigned groups it would indicate that our analysis was likely to be picking up statistical artifacts in addition to true affects.  Of course there is a trade off between sensitivity and specificity and a small number of false positives may be acceptable in order to find more true positives.  

When we compared the female-male results it was found that there were actually more genes found with an unadjusted p-value of less than 0.05 in the randomised groups compared to the female-male groups, 1040 compared to 864. When using an unadjusted p-value of 0.05 we did expect to find a large number of genes since there are 44061 genes in total however it is a bit surprising that there was a greater number of genes found in the random grouping.  However once we we used the corrected p-value taking into acount the number of tests performed the number of genes found in the random grouping went down to 1 compared to 61 which is a good result giving us confidence that after adjustment we will be getting a low number of false positives.  We can see this as well in the top ten list of genes.  The randomised group does tend to a have higher unadjusted p-values although they are much lower than the significance level however the main difference is the false discovery rate which are much higher in the random grouping.

The heatmap is also interesting because we can see that there is no clear separation between the groups as there was in the female-male groups and in fact we can see that a single individual with very high expression can be enough to create the differential expression by itself.  

There were no Y chromosome genes found which was to be expected since the males and females were mixed together in the groupings and so we would expect those genes to be approximately equal between the random groups.  Similarly the XIST gene only had a -0.4047 log fold change between the groups which is quite a small number.